src/hooks/useBalance.ts

```diff
@@ -45,34 +45,29 @@ const useBalance = (api: ApiPromise, address: string, providedSi: boolean = fals
   const [state, setState] = useMountedState<State>(initValues);
 
   useEffect((): (() => void) => {
-    const getBalance = async (api: ApiPromise, address: string, setState: any): Promise<VoidFn> => {
-      try {
-        const u = await api.query.system.account(address, ({ data }): void => {
-          setState({
-            chainTokens: data.free.registry.chainTokens[0],
-            formattedBalance: formatBalance(data.free, {
-              decimals: api.registry.chainDecimals[0],
-              forceUnit: '-',
-              withSi: providedSi
-            }),
-            free: data.free
-          });
+    let unsubscribe: null | (() => void) = null;
+    api?.query?.system
+      .account(address, ({ data }): void => {
+        setState({
+          chainTokens: data.free.registry.chainTokens[0],
+          formattedBalance: formatBalance(data.free, {
+            decimals: api.registry.chainDecimals[0],
+            forceUnit: '-',
+            withSi: providedSi
+          }),
+          free: data.free
         });
-        return Promise.resolve(u);
-      } catch (e) {
+      })
+      .then((u): void => {
+        unsubscribe = u;
```

Comment:
You have just replaced one race condition into another. IF the component is unmounted before we hit this line we will never unsubscribe.

Reply:
If the component is unmounted then I do not understand how this line is relevant to unsubscription...?

Reply:
1. We mount the component.
2. `useEffect` fires and we start with `api.query.system.account`
3. The promise is not resolved yet (i.e. we didn't invoke `then` yet).
4. The component is unmounted and cleanup runs, but since we didn't reach `then` yet, `unsubscribe === null` so we don't unsubscribe from anything.
5. The promise resolves and we set `unsubscribe = u` - the subscription is now estabilisihed, but the component has already been unmounted, so every `setState` will fail because `isMounted() === false`. Also see #141 for something that would allow us to detect some problems.

Now we have a leak, which is really hard to find, but esentially the subscription will be there during entire lifetime of application.

Reply:
You are right on this. I have though found the leak and roll back also the code to the previous one.
Have a look on #144